### ISSUE TITLE
Rule Change: (fill in)

### ISSUE THREAD

**Issue by rfool** [eslint/eslint#15932]
### What rule do you want to change?

space-unary-ops

### What change to do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

Other

### Example code

```js
// new: space required
const f = new foo;

// typeof: we dont care
typeof(a) === 'number';
typeof a === 'number';

// nonwords ops: we dont care
let a = -23;
let b = - 23;
++a;
a ++;
a++;
--b;
-- b;
b--;
b --;
```

### What does the rule currently do for this code?

The current implementation of rule _will always apply_ both its `words` and `nonwords` options.

The `overrides` option, when used, _will always apply_ too.

But there is **no way to relax (parts of) the rule**.

For example, one would like to apply this rule to all `word` ops, but also permit multiple layouts for the `typeof` op. As in example code above.

Or another case: one wants to apply the rule to all `word` ops, but leave the `nonword` ops untouched.

### What will the rule do after it's changed?

**My proposal:**

Make it possible to have **three states** for the options: `true`, `false` and `null` (new). The rule ignores appropriate cases when an option is `null`.

The `overrides` option may also contain `null` entries - these would effectively disable appropriate cases.

**Example rule config:**
```js
'space-unary-ops': [
  'error',
  {
    words: true,
    nonwords: false,
    overrides: {
      '++': null,      // <- with/without space - both is valid
      '--': null,
      '-': null,
      'typeof': null,
    }
  }
],
```
---

The meta data schema of options would change to:
```js
schema: [
  {
    type: "object",
    properties: {
      words: {
        type: [ "boolean", "null" ],
        default: true
      },
      nonwords: {
        type: [ "boolean", "null" ],
        default: false
      },
      overrides: {
        type: "object",
        additionalProperties: {
          type: [ "boolean", "null" ],
        }
      }
    },
    additionalProperties: false
  }
],
```

---

The code changes are rather trivial - just a third condition to check in a few places -  for example:

```js
function checkUnaryWordOperatorForSpaces(node, firstToken, secondToken, word) {
  if (overrideExistsForOperator(word)) {
    const overrideWords = overrideEnforcesSpaces(word);
    if (overrideWords===true) {
      verifyWordHasSpaces(node, firstToken, secondToken, word);
    } else if (overrideWords===false) {
      verifyWordDoesntHaveSpaces(node, firstToken, secondToken, word);
    } else {
      // skip verification when override.words===null
    }
  } else if (options.words===true) {
    verifyWordHasSpaces(node, firstToken, secondToken, word);
  } else if (options.nonwords===false) {
    verifyWordDoesntHaveSpaces(node, firstToken, secondToken, word);
  } else {
    // skip verification when options.words===null
  }
}
```

### Participation

- [X] I am willing to submit a pull request to implement this change.

### Additional comments

**Generally ?**

There are other rules which tie things together in a way that makes them very unflexible.

Another example: `object-curly-newline` will always apply to all four sub-types (ObjectExpression, ObjectPattern, ImportDeclaration, ExportDeclaration). Though it is nice, that each type can have different options, it is bad that one can not have rules that apply to one type while ignoring the other types. It is always all-in - all or nothing. Here, a `null` option could be useful too.

There are probably many other more rules that could be improved by such relaxed options.

**Comment by mdjermanovic** [comment]
Hi @rfool, thanks for the issue!

Per our rules policies, [stylistic rules are frozen](https://github.com/eslint/eslint#stylistic-rule-updates), so we are not adding new options (including new values for existing options) to the `space-unary-ops` rule. You can always copy the existing rule into a custom rule and make it behave the way you want.
